engine/src/main/java/org/terasology/rendering/world/WorldRendererImpl.java

```diff
@@ -249,15 +252,10 @@ private void initRenderGraph() {
         renderGraph.add(bloomPassesNode);
         renderGraph.add(blurPassesNode);
         renderGraph.add(finalPostProcessingNode);
-    }
 
-    private static <T extends Node> T createInstance(Class<T> type, Context context) {
-        // Attempt constructor-based injection first
-        T node = InjectionHelper.createWithConstructorInjection(type, context);
-        // Then fill @In fields
-        InjectionHelper.inject(node, context);
-        node.initialise();
-        return type.cast(node);
+        RenderPipelineGenerator generator = new RenderPipelineGenerator(renderGraph);
+        RenderPipelineOptimizer optimizer = new RenderPipelineOptimizer(generator.generate());
+        renderPipelineExecutor = new RenderPipelineExecutor(optimizer.optimize());
```

Comment:
Hmmm... I think you are right. Let's wrap all the complexity into a single class that _internally_ goes through the stages of generating the list of Nodes+StateChanges, optimizes it by pruning redundant StateChanges and then generates a task list as output.

I'd prefer to stick to 

`RenderTaskListGenerator.createFrom(orderedNodes)`

instead of passing the whole render graph in. One easy reason for this is that you'd normally pass a whole instance if you need to call more than one method from it, which here isn't the case. Furthermore, at some point we might want the render graph to output multiple lists of nodes, to be evaluated independently for performance. Having a createFrom(listOfNodes)  method will mean just calling it multiple times with different lists. Having a createFrom(renderGraph) method would mean you'd have to change something in it when the render graph can output multiple lists.